src/txmempool.cpp

```diff
@@ -150,33 +151,48 @@ void CTxMemPool::UpdateTransactionsFromBlock(const std::vector<uint256> &vHashes
+bool CTxMemPool::CalculateMemPoolAncestors(const std::vector<CTxMemPoolEntry>& entries,
+                                           setEntries& setAncestors,
+                                           const uint64_t limitAncestorCount,
+                                           const uint64_t limitAncestorSize,
+                                           const uint64_t limitDescendantCount,
+                                           const uint64_t limitDescendantSize,
+                                           std::string &errString,
+                                           const bool fSearchForParents /*=true*/) const
 {
-    CTxMemPoolEntry::Parents staged_ancestors;
-    const CTransaction &tx = entry.GetTx();
+    const size_t total_virtual_size = std::accumulate(entries.cbegin(), entries.cend(), 0,
+                                      [](int64_t sum, const auto& entry)
+                                      { return sum + GetVirtualTransactionSize(entry.GetTx()); });
+    const size_t total_count = entries.size();
 
+    CTxMemPoolEntry::Parents staged_ancestors;
     if (fSearchForParents) {
-        // Get parents of this transaction that are in the mempool
-        // GetMemPoolParents() is only valid for entries in the mempool, so we
-        // iterate mapTx to find parents.
-        for (unsigned int i = 0; i < tx.vin.size(); i++) {
-            std::optional<txiter> piter = GetIter(tx.vin[i].prevout.hash);
-            if (piter) {
-                staged_ancestors.insert(**piter);
-                if (staged_ancestors.size() + 1 > limitAncestorCount) {
-                    errString = strprintf("too many unconfirmed parents [limit: %u]", limitAncestorCount);
-                    return false;
+
+        for (const auto& entry : entries) {
+            // Get parents of this transaction that are in the mempool
+            // GetMemPoolParents() is only valid for entries in the mempool, so we
+            // iterate mapTx to find parents.
+            for (const auto& input : entry.GetTx().vin) {
+                std::optional<txiter> piter = GetIter(input.prevout.hash);
+                if (piter) {
+                    staged_ancestors.insert(**piter);
+                    if (staged_ancestors.size() + 1 > limitAncestorCount) {
+                        errString = strprintf("too many unconfirmed parents [limit: %u]", limitAncestorCount);
+                        return false;
+                    }
                 }
             }
         }
     } else {
-        // If we're not searching for parents, we require this to be an
-        // entry in the mempool already.
-        txiter it = mapTx.iterator_to(entry);
-        staged_ancestors = it->GetMemPoolParentsConst();
+        // If we're not searching for parents, we require all entries to be in the mempool already.
+        for (const auto& entry : entries) {
```

Comment:
If you're touching this code you *could* update it to use Epoch Algorithms.

Essentially all you need to do is get a new epoch, and then visit the txiters you are interested in and add them to a vec instead of a set.

This lets you get rid of the O(n^2)  potential of a merge call.

Reply:
Alternatively, I can PR the epoch algorithm and you can rebase on it, or I can PR it before this code gets released.

Reply:
> Alternatively, I can PR the epoch algorithm and you can rebase on it, or I can PR it before this code gets released.

Both sound like good options!